### ISSUE TITLE
Help with Windows Store Certificates

### ISSUE THREAD

**Issue by amaljesudas** [openssl/openssl#17924]
Hi,

Need to implement TLS communication on an application that has the potential to be a client as well as server.
For certificate setting, need to do via memory instead of physical files or paths.
So if I have installed my certificate in my Windows Certificate Store, I need to use the same to set my public and private key.
My environment is C++ with openssl 1.1.1l.
I could do this for the public key but am stuck with accessing the private key.
Have tried many possibilities, but none seem to work.
Can some one guide me on how I could extract the private key of a certificate from Windows Certificate Store?
Also, some input on the openssl API that would allow me to set the private key from memory.

Thanks in advance.

Regards,
Amal

**Comment by bernd-edlinger** [comment]
By default You can't extract the private key.

However, using the capi engine you can use the private key to
create RSA and DSA signatures,
and decode PKCS1_PADDING and NO_PADDING RSA encrypted data.

That should be sufficient to make TLS work without direct access to
the private key.

Note that the capi engine is broken in openssl-3.0 and will not work there.

But there is probably also a way to mark a private key as exportable.

Coincidentally I know of code that claims to be able to export a private key
in the OPC/UA LDS application which might provide an alternative solution:
https://github.com/OPCFoundation/UA-LDS/blob/6a1faa1973afe5f0c90c2969f0312ad970ccaee7/stack/Stack/platforms/win32/opcua_p_win32_pki.c#L556-L842
The Comment here says:
```
 * note: for this to work the Certificate + Private Key must be in PKCS#12 format
 *       and imported into the Store "MY" / "Eigene Zertifikate"
 * important: please check "Schluessel als exportierbar markieren"
 * openssl pkcs12 -export -in cert.pem -inkey private_key.pem -out cert.p12
 * Internet -> Inhalte -> Zertifikate... -> Eigene Zertifikate -> Importieren...
```
So, as far as I recall you need to set that flag when you import the
certificate together with the private key into the windows store.
But I don't know if that still works this way with current windows versions.

**Comment by amaljesudas** [comment]
@bernd-edlinger 

Thanks Bernd for all the information.
I don't think Openssl in our environment was built enabling capi engine. 
Will have to rely on WinCrypt and common openssl APIs.
The code you shared looks promising, will definitely try it out.
And, yes I have set the flag that marks private key as exportable.

Thanks & Regards,
Amal

**Comment by amaljesudas** [comment]
@bernd-edlinger 

I have succeeded in using the private key and certificate from the  windows store without saving as physical files.
The sample code shared by you gave enough clues. Thanks for providing the same.
Also, capi engine support was there in our environment. I thought we needed to build again with specific configuration. My bad.
Any way, what I was looking for is now done. Thanks a ton. 

Thanks & Regards,
Amal
